### DISCUSSION TITLE
I upgraded MailKit from v2.x and now the AddFlags() method doesn't seem to exist?

### DISCUSSION THREAD

**Discussion by Stephan-4711** [jstedfast/MailKit#1646] [Q&A] [answered]
Hi,

I was using an old Version Mail and MimeKit now I switched from Windows to Linux an used the newest version to install.

The AddFlags seems to work diffrent now? 
`#$imap.Inbox.AddFlags($id,[MailKit.MessageFlags]::Seen,$true)
$imap.Inbox.AddFlags($id,[MailKit.MessageFlags]::Deleted)
$imap.Inbox.Expunge`

The marked messages as seen is working on my Windows System.
on my newest Linux System with powershell I get the following error:
| Method invocation failed because [MimeKit.MimeMessage] does not contain a method named 'AddFlags'.

The delete message part I was also trying to add instead of mark the messages as seen, but this also never worked for me.
Anyone can help me?

**Comment by jstedfast** [comment, marked as answer]
Hi @Stephan-4711,

A lot of API changes were made in [MailKit v3.0](https://github.com/jstedfast/MailKit/blob/master/ReleaseNotes.md#mailkit-300-2021-12-11) to combat the explosion of ImapFolder overloads that were being added in order to support additional IMAP extensions that MailKit was being updated to support.

In v2.x for example, MailKit had the following AddFlags() method overloads:

* `AddFlags(int index, MessageFlags flags, bool silent, CancellationToken cancellationToken)`
* `AddFlags(UniqueId uid, MessageFlags flags, bool silent, CancellationToken cancellationToken)`
* `AddFlags(IList<int> indexes, MessageFlags flags, bool silent, CancellationToken cancellationToken)`
* `AddFlags(IList<UniqueId> uid, MessageFlags flags, bool silent, CancellationToken cancellationToken)`
* `AddFlags(int index, MessageFlags flags, HashSet<string> keywords, bool silent, CancellationToken cancellationToken)`
* `AddFlags(UniqueId uid, MessageFlags flags, HashSet<string> keywords, bool silent, CancellationToken cancellationToken)`
* `AddFlags(IList<int> indexes, MessageFlags flags, HashSet<string> keywords, bool silent, CancellationToken cancellationToken)`
* `AddFlags(IList<UniqueId> uid, MessageFlags flags, HashSet<string> keywords, bool silent, CancellationToken cancellationToken)`
* `AddFlags(int index, MessageFlags flags, long modseq, bool silent, CancellationToken cancellationToken)`
* `AddFlags(UniqueId uid, MessageFlags flags, long modseq, bool silent, CancellationToken cancellationToken)`
* `AddFlags(IList<int> indexes, MessageFlags flags, long modseq, bool silent, CancellationToken cancellationToken)`
* `AddFlags(IList<UniqueId> uid, MessageFlags flags, long modseq, bool silent, CancellationToken cancellationToken)`
* `AddFlags(int index, MessageFlags flags, HashSet<string> keywords, long modseq, bool silent, CancellationToken cancellationToken)`
* `AddFlags(UniqueId uid, MessageFlags flags, HashSet<string> keywords, long modseq, bool silent, CancellationToken cancellationToken)`
* `AddFlags(IList<int> indexes, MessageFlags flags, HashSet<string> keywords, long modseq, bool silent, CancellationToken cancellationToken)`
* `AddFlags(IList<UniqueId> uid, MessageFlags flags, HashSet<string> keywords, long modseq, bool silent, CancellationToken cancellationToken)`
* plus the Async equivalents plus RemoveFlags/Async() equivalents plus SetFlags/Async() equivalents ...and probably more! And that was just for the APIs to update the message flags!

It was just getting way too out of hand for me to maintain, so for v3.0, I made all of those into extensions methods located in `IMailFolderExtensions` and replaced them with a much smaller set of `Store()` overloads that take an IStoreRequest that define all of the options.

To resolve the problem you are hitting, there are 2 options:

1. Call the [IMailFolderExtensions](http://mimekit.net/docs/html/T_MailKit_IMailFolderExtensions.htm) AddFlags() method
2. Create a [StoreFlagsRequest](http://mimekit.net/docs/html/T_MailKit_StoreFlagsRequest.htm) object and then call `$imap.Inbox.Store($id, $request)`

BTW, to reduce network traffic, you can set *both* the Seen and Deleted flags in a single AddFlags or Store request. You don't need to make 2 calls. The MessageFlags enum is a set of `[Flags]` which means that you can bitwise-or them together.

**Reply by Stephan-4711** [reply]
Thanks, sadly I'm absoluty to stupid for programming.
I'll ask a coworker to help me.
I just can push mice and administation of Windows Servers ;-)
